firedrake/preconditioners/pmg.py

```diff
@@ -482,400 +495,583 @@ def coarsen_bc_value(self, bc, cV):
         return coarse
 
 
-def prolongation_transfer_kernel_aij(Pk, P1):
-    # Works for Pk, Pm; I just retain the notation
-    # P1 to remind you that P1 is of lower degree
-    # than Pk
+def prolongation_transfer_kernel_action(Vf, expr):
     from tsfc import compile_expression_dual_evaluation
     from tsfc.finatinterface import create_element
-    from firedrake import TestFunction
+    to_element = create_element(Vf.ufl_element())
+    kernel = compile_expression_dual_evaluation(expr, to_element, Vf.ufl_element())
+    coefficients = kernel.coefficients
+    if kernel.first_coefficient_fake_coords:
+        target_mesh = Vf.ufl_domain()
+        coefficients[0] = target_mesh.coordinates
 
-    expr = TestFunction(P1)
-    to_element = create_element(Pk.ufl_element())
+    return op2.Kernel(kernel.ast, kernel.name,
+                      requires_zeroed_output_arguments=True,
+                      flop_count=kernel.flop_count), coefficients
 
-    kernel = compile_expression_dual_evaluation(expr, to_element, Pk.ufl_element())
-    ast = kernel.ast
-    name = kernel.name
-    flop_count = kernel.flop_count
-    return op2.Kernel(ast, name, requires_zeroed_output_arguments=True,
-                      flop_count=flop_count)
 
-
-def tensor_product_space_query(V):
+@lru_cache(maxsize=10)
+def expand_element(ele):
+    """
+    Expand a FiniteElement as an EnrichedElement of TensorProductElements, discarding modifiers.
```

Comment:
I guess the goal here is to get a "scalar" element with nodes in the same place as the incoming element?

Reply:
AKA: can you expand a little more in this docstring.